File: 杰理之家/【杰理之家】SDK_v1.6.0】/NewJieliZhiNeng/NewJieliZhiNeng/设备/设备信息/DhaFittingView/DhaChartView/EcChartsView.m
//
//  EcChartsView.m
//  NewJieliZhiNeng
//
//  Created by EzioChan on 2022/7/4.
//  Copyright © 2022 杰理科技. All rights reserved.
//

#import "EcChartsView.h"
#import <Charts/Charts-Swift.h>

@interface EcChartsView()<ChartViewDelegate,IChartAxisValueFormatter>{
    CombinedChartView *cbView;
}
@end

@implementation EcChartsView

-(instancetype)initWithFrame:(CGRect)frame{
    self = [super initWithFrame:frame];
    if (self) {
        cbView = [CombinedChartView new];
        [self addSubview:cbView];
        [cbView mas_makeConstraints:^(MASConstraintMaker *make) {
            make.top.equalTo(self.mas_top).offset(0);
            make.left.equalTo(self.mas_left).offset(0);
            make.right.equalTo(self.mas_right).offset(0);
            make.bottom.equalTo(self.mas_bottom).offset(0);
        }];
        self.maxIndex = 6;
        self.shouldTag = -1;
        
        cbView.delegate = self;
        cbView.drawOrder = @[@(CombinedChartDrawOrderBar),@(CombinedChartDrawOrderLine)];
        cbView.noDataText = @"暂无数据"; // 无数据时显示的文字
        cbView.pinchZoomEnabled = NO; // 触控放大
        cbView.doubleTapToZoomEnabled = NO; // 双击放大
        cbView.scaleXEnabled = NO; // X 轴缩放
        cbView.scaleYEnabled = NO; // Y 轴缩放
        cbView.scaleEnabled = NO; // 缩放
        cbView.highlightPerTapEnabled = NO; // 单击高亮
        cbView.highlightPerDragEnabled = NO; // 拖拽高亮
        cbView.dragEnabled = NO; // 拖拽图表
        cbView.dragDecelerationEnabled = YES; // 拖拽后是否有惯性效果
        cbView.dragDecelerationFrictionCoef = 0.5; // 拖拽后惯性效果的摩擦系数(0~1)，数值越小，惯性越不明显
        cbView.legend.enabled = NO; // 隐藏图例
        
        
        /* 设置 X 轴显示的值的属性 */
        ChartXAxis *xAxis = cbView.xAxis;
        xAxis.labelPosition = XAxisLabelPositionBottom; // 显示位置
        xAxis.drawGridLinesEnabled = NO; // 网格绘制
        xAxis.axisLineColor = [UIColor colorFromHexString:@"#ABABAB"]; // X 轴颜色
        xAxis.axisLineWidth = 0.5f; // X 轴线宽
        xAxis.labelFont = [UIFont systemFontOfSize:13]; // 字号
        xAxis.labelTextColor = [UIColor colorFromHexString:@"#ABABAB"]; // 颜色
        xAxis.axisBackgroundColor = [UIColor colorFromHexString:@"#805BEB"];
        xAxis.labelRotationAngle = 0; // 文字倾斜角度
        xAxis.centerAxisLabelsEnabled = YES;
        xAxis.labelCount = 7;
        xAxis.granularity = 1.0;
        xAxis.axisMinimum = 0; // X轴最小值
        xAxis.axisMaximum = self.maxIndex;
        xAxis.valueFormatter = self;
        xAxis.axisLineColor = [UIColor clearColor];
        
        cbView.rightAxis.drawLabelsEnabled = false;
        cbView.rightAxis.drawGridLinesEnabled = NO;
        cbView.rightAxis.enabled = NO;
        
        ChartYAxis *leftAxis = cbView.leftAxis;
        leftAxis.labelPosition = YAxisLabelPositionOutsideChart; // 显示位置
        leftAxis.drawGridLinesEnabled = YES; // 网格绘制
        leftAxis.gridColor = [UIColor colorFromHexString:@"#ABABAB"]; // 网格颜色
        leftAxis.gridLineDashLengths = @[@2,@4];
        leftAxis.labelFont = [UIFont systemFontOfSize:13]; // 字号
        leftAxis.labelTextColor = [UIColor colorFromHexString:@"#ABABAB"]; // 颜色
        leftAxis.axisMinimum = 0; // 最小值
        leftAxis.axisMaximum = 100; // 最大值（不设置会根据数据自动设置）
        leftAxis.decimals = 2;
        leftAxis.axisLineColor = [UIColor clearColor];
        
        }
    return self;
}


-(void)setBarValues:(NSArray *)barValues{
    _barValues = barValues;
    
    NSMutableArray<BarChartDataEntry *> *entries1 = [[NSMutableArray alloc] init];
    NSMutableArray<BarChartDataEntry *> *entries2 = [[NSMutableArray alloc] init];
    for (int index = 0; index < barValues.count; index++)
    {
        [entries1 addObject:[[BarChartDataEntry alloc] initWithX:index+0.5 y:100.0]];
        
        NSNumber *number = barValues[index];
        if (self.shouldTag == index) {
            [entries2 addObject:[[BarChartDataEntry alloc] initWithX:index+0.5 y:[number floatValue] icon:[UIImage imageNamed:@"Theme.bundle/icon_dha_tips.png"]]];
        }else{
            [entries2 addObject:[[BarChartDataEntry alloc] initWithX:index+0.5 y:[number floatValue]]];
        }
        
    }
    BarChartDataSet *dataSet = [[BarChartDataSet alloc] initWithEntries:entries1];
    dataSet.barGradientColors = @[@[[UIColor colorFromHexString:@"#E9EBF0"],[UIColor colorFromHexString:@"#E9EBF0"]]];
    dataSet.axisDependency = AxisDependencyLeft; // 根据左边数据显示
    dataSet.drawValuesEnabled = NO; // 是否显示数据
    
    
    BarChartDataSet *dataSet2 = [[BarChartDataSet alloc] initWithEntries:entries2];
    dataSet2.barGradientColors = @[@[[UIColor colorFromHexString:@"#A896DD"],[UIColor colorFromHexString:@"#805BEB"]]];
    dataSet2.axisDependency = AxisDependencyLeft; // 根据左边数据显示
    dataSet2.drawValuesEnabled = NO; // 是否显示数据
    dataSet2.iconsOffset = CGPointMake(0, 13);
    
    BarChartData *data = [[BarChartData alloc] initWithDataSets:@[dataSet,dataSet2]];
    data.barWidth = 0.65f; // 柱状图宽度（数值范围 0 ~ 1）
    
    CombinedChartData *charData = [[CombinedChartData alloc] init];
    charData.barData = data;
    charData.lineData = nil;
    cbView.data = charData;
    [cbView notifyDataSetChanged];
    
}

-(void)setShouldTag:(NSInteger)shouldTag{
    _shouldTag = shouldTag;
    cbView.xAxis.axisHaveBgC = @[@(shouldTag+1)];
}


-(void)setLineArrays:(NSArray * _Nullable )arr1 Array2:(NSArray  * _Nullable )arr2{
    NSMutableArray<BarChartDataEntry *> *entries1 = [[NSMutableArray alloc] init];
    for (int index = 0; index < self.xAxisList.count; index++)
    {
        [entries1 addObject:[[BarChartDataEntry alloc] initWithX:index+0.5 y:100.0]];
    }
    BarChartDataSet *dataSet = [[BarChartDataSet alloc] initWithEntries:entries1 label:@"bgBar"];
    dataSet.barGradientColors = @[@[[UIColor colorFromHexString:@"#E9EBF0"],[UIColor colorFromHexString:@"#E9EBF0"]]];
    dataSet.axisDependency = AxisDependencyLeft; // 根据左边数据显示
    dataSet.drawValuesEnabled = NO; // 是否显示数据
    BarChartData *data = [[BarChartData alloc] initWithDataSets:@[dataSet]];
    data.barWidth = 0.65; // 柱状图宽度（数值范围 0 ~ 1）
    
    LineChartData *d1 = [[LineChartData alloc] init];
    
    
    NSMutableArray *entries = [[NSMutableArray alloc] init];
    NSMutableArray *entries2 = [[NSMutableArray alloc] init];
    
    if (arr1) {
        for (int index = 0; index < arr1.count; index++)
        {
            NSNumber *number = arr1[index];
            [entries addObject:[[ChartDataEntry alloc] initWithX:index + 0.5 y:[number floatValue]]];
        }
        LineChartDataSet *set = [[LineChartDataSet alloc] initWithEntries:entries label:@"Line DataSet"];
        [set setColor:[UIColor colorFromHexString:@"#8DB3FA"]];
        set.lineWidth = 2.0;
        [set setCircleColor:[UIColor colorFromHexString:@"#4E89F4"]];
        set.circleRadius = 4.0;
        set.circleHoleRadius = 0;
        set.fillColor = [UIColor colorFromHexString:@"#4E89F4"];
        set.mode = LineChartModeLinear;
        set.drawValuesEnabled = NO;
        [d1 addDataSet:set];
    }
    
    if (arr2) {
        for (int index = 0; index < arr2.count; index++)
        {
            NSNumber *number = arr2[index];
            [entries2 addObject:[[ChartDataEntry alloc] initWithX:index + 0.5 y:[number floatValue]]];
        }
        
        LineChartDataSet *set2 = [[LineChartDataSet alloc] initWithEntries:entries2 label:@"Line DataSet2"];
        [set2 setColor:[UIColor colorFromHexString:@"#F7C189"]];
        set2.lineWidth = 2.0;
        [set2 setCircleColor:[UIColor colorFromHexString:@"#FF9E39"]];
        set2.circleRadius = 4.0;
        set2.circleHoleRadius = 0;
        set2.fillColor = [UIColor colorFromHexString:@"#FF9E39"];
        set2.mode = LineChartModeLinear;
        set2.drawValuesEnabled = NO;
        [d1 addDataSet:set2];
    }
    
    
    CombinedChartData *charData = [[CombinedChartData alloc] init];
    charData.barData = data;
    charData.lineData = d1;
    cbView.data = charData;
    [cbView notifyDataSetChanged];
}





- (void)chartValueSelected:(ChartViewBase * _Nonnull)chartView entry:(ChartDataEntry * _Nonnull)entry highlight:(ChartHighlight * _Nonnull)highlight{
    
}
/// Called when a user stops panning between values on the chart
- (void)chartViewDidEndPanning:(ChartViewBase * _Nonnull)chartView{
    
}
- (void)chartValueNothingSelected:(ChartViewBase * _Nonnull)chartView{
    
}
- (void)chartScaled:(ChartViewBase * _Nonnull)chartView scaleX:(CGFloat)scaleX scaleY:(CGFloat)scaleY{
    
}
- (void)chartTranslated:(ChartViewBase * _Nonnull)chartView dX:(CGFloat)dX dY:(CGFloat)dY{
    
}
- (void)chartView:(ChartViewBase * _Nonnull)chartView animatorDidStop:(ChartAnimator * _Nonnull)animator{
    
}


-(NSString *)stringForValue:(double)value axis:(ChartAxisBase *)axis{
    return self.xAxisList[(int)value%self.xAxisList.count];
}


@end

File: code/JieLi_Home_Demo/NewJieliZhiNeng/DevicesViewController/DeviceInfoVC/DhaFittingVC/DhaChartView/EcChartsView.swift
import UIKit
import DGCharts
import SnapKit

@objcMembers
class EcChartsView: UIView {

    // MARK: - Public properties (ObjC 兼容)
    var xAxisList: [String] = [] {
        didSet { updateXAxisFormatter() }
    }

    var maxIndex: Int = 6 {
        didSet { updateXAxisRange() }
    }

    // 高亮的 x 轴索引（用于在柱状条上展示 icon 提示）
    var shouldTag: Int = -1 {
        didSet { refreshBarIcons() }
    }

    // 外部设置的柱状图数据（第二组真实值，第一组为背景填充）
    var barValues: [NSNumber] = [] {
        didSet { setBarValues(barValues) }
    }

    // MARK: - Private
    private let combinedView = CombinedChartView()

    // 记录当前的组合数据，以便局部刷新（比如只更新 icon）
    private var currentBarDataSetBg: BarChartDataSet?
    private var currentBarDataSetVal: BarChartDataSet?

    // MARK: - Init
    override init(frame: CGRect) {
        super.init(frame: frame)
        setupChart()
    }

    required init?(coder: NSCoder) {
        super.init(coder: coder)
        setupChart()
    }

    // MARK: - Public API (保持 ObjC 调用兼容的 selector)
    // ObjC 原型: - (void)setLineArrays:(NSArray * _Nullable)arr1 Array2:(NSArray * _Nullable)arr2;
    @objc(setLineArrays:Array2:)
    func setLineArrays(_ arr1: [NSNumber]?, Array2 arr2: [NSNumber]?) {
        let count = max(xAxisList.count, maxIndex + 1)
        var entries1: [ChartDataEntry] = []
        var entries2: [ChartDataEntry] = []

        // 折线1
        if let a1 = arr1 {
            for i in 0..<count {
                let y = i < a1.count ? a1[i].doubleValue : 0.0
                entries1.append(ChartDataEntry(x: Double(i), y: y))
            }
        }

        // 折线2
        if let a2 = arr2 {
            for i in 0..<count {
                let y = i < a2.count ? a2[i].doubleValue : 0.0
                entries2.append(ChartDataEntry(x: Double(i), y: y))
            }
        }

        var lineDataSets: [LineChartDataSet] = []
        if !entries1.isEmpty {
            let ds1 = LineChartDataSet(entries: entries1, label: "")
            ds1.axisDependency = YAxis.AxisDependency.left
            ds1.setColor(UIColor.white)
            ds1.lineWidth = 3
            ds1.circleRadius = 0
            ds1.drawCirclesEnabled = false
            ds1.drawValuesEnabled = false
            lineDataSets.append(ds1)
        }

        if !entries2.isEmpty {
            let ds2 = LineChartDataSet(entries: entries2, label: "")
            ds2.axisDependency = YAxis.AxisDependency.left
            ds2.setColor(hex("#67D0F0"))
            ds2.lineWidth = 3
            ds2.circleRadius = 0
            ds2.drawCirclesEnabled = false
            ds2.drawValuesEnabled = false
            lineDataSets.append(ds2)
        }

        let combined = combinedView.data as? CombinedChartData ?? CombinedChartData()
        combined.lineData = lineDataSets.isEmpty ? nil : LineChartData(dataSets: lineDataSets)
        combinedView.data = combined
        combinedView.notifyDataSetChanged()
    }

    // MARK: - Private setup
    private func setupChart() {
        addSubview(combinedView)
        combinedView.snp.makeConstraints { make in
            make.edges.equalToSuperview()
        }

        // 交互设置（与 ObjC 版本保持一致）
        combinedView.delegate = self
        combinedView.drawOrder = [CombinedChartView.DrawOrder.bar.rawValue,
                                   CombinedChartView.DrawOrder.line.rawValue]
        combinedView.noDataText = ""
        combinedView.chartDescription.enabled = false
        combinedView.pinchZoomEnabled = false
        combinedView.doubleTapToZoomEnabled = false
        combinedView.scaleXEnabled = false
        combinedView.scaleYEnabled = false
        combinedView.dragEnabled = false
        combinedView.highlightPerTapEnabled = false
        combinedView.highlightPerDragEnabled = false
        combinedView.dragDecelerationEnabled = false
        combinedView.dragDecelerationFrictionCoef = CGFloat(0x10) / 100.0

        combinedView.legend.enabled = false

        // X 轴
        let xAxis = combinedView.xAxis
        xAxis.labelPosition = .bottom
        xAxis.labelFont = .systemFont(ofSize: 14)
        xAxis.labelTextColor = hex("#FFFFFF")
        xAxis.axisLineColor = .clear // 不显示 x 轴线
        xAxis.gridColor = .clear
        xAxis.centerAxisLabelsEnabled = true
        xAxis.granularity = 1
        xAxis.labelRotationAngle = 0

        // 左轴
        let left = combinedView.leftAxis
        left.labelTextColor = hex("#FFFFFF")
        left.axisLineColor = .clear
        left.gridColor = hex("#545454")
        left.gridLineDashLengths = [2, 4]
        left.axisMinimum = 0

        // 右轴不使用
        combinedView.rightAxis.enabled = false

        // 初始范围与 formatter
        updateXAxisRange()
        updateXAxisFormatter()
    }

    private func updateXAxisRange() {
        let xAxis = combinedView.xAxis
        xAxis.axisMinimum = 0
        xAxis.axisMaximum = Double(maxIndex)
        xAxis.labelCount = maxIndex + 1
        combinedView.notifyDataSetChanged()
    }

    private func updateXAxisFormatter() {
        combinedView.xAxis.valueFormatter = IndexAxisValueFormatter(values: xAxisList)
        combinedView.notifyDataSetChanged()
    }

    private func refreshBarIcons() {
        // 仅更新柱状条的 icon 标记
        guard let dsVal = currentBarDataSetVal else { return }
        let count = dsVal.entries.count
        for i in 0..<count {
            if i == shouldTag {
                dsVal.entries[i].icon = UIImage(named: "Theme.bundle/icon_dha_tips.png")
            } else {
                dsVal.entries[i].icon = nil
            }
        }
        dsVal.iconsOffset = CGPoint(x: 0, y: 18)
        combinedView.data?.notifyDataChanged()
        combinedView.notifyDataSetChanged()
    }

    // 与 ObjC 的 setBarValues 行为一致：
    // - 背景柱状条固定为 100（用于背景填充）
    // - 实际数值柱状条来自传入的数组
    private func setBarValues(_ values: [NSNumber]) {
        let count = max(xAxisList.count, maxIndex + 1)
        var entriesBg: [BarChartDataEntry] = []
        var entriesVal: [BarChartDataEntry] = []

        for i in 0..<count {
            entriesBg.append(BarChartDataEntry(x: Double(i), y: 100.0))
            let y = i < values.count ? values[i].doubleValue : 0.0
            let entry = BarChartDataEntry(x: Double(i), y: y)
            if i == shouldTag {
                entry.icon = UIImage(named: "Theme.bundle/icon_dha_tips.png")
            }
            entriesVal.append(entry)
        }

        let dsBg = BarChartDataSet(entries: entriesBg, label: "")
        dsBg.axisDependency = YAxis.AxisDependency.left
        dsBg.colors = [hex("#2E92DA")]
        dsBg.drawValuesEnabled = false

        let dsVal = BarChartDataSet(entries: entriesVal, label: "")
        dsVal.axisDependency = YAxis.AxisDependency.left
        dsVal.colors = [hex("#67D0F0")]
        dsVal.drawValuesEnabled = false
        dsVal.iconsOffset = CGPoint(x: 0, y: 18)

        currentBarDataSetBg = dsBg
        currentBarDataSetVal = dsVal

        let barData = BarChartData(dataSets: [dsBg, dsVal])
        barData.barWidth = 0.4

        let combined = combinedView.data as? CombinedChartData ?? CombinedChartData()
        combined.barData = barData
        combinedView.data = combined
        combinedView.notifyDataSetChanged()
    }

    // MARK: - Helpers
    private func hex(_ string: String) -> UIColor {
        var str = string.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
        if str.hasPrefix("#") { str.removeFirst() }
        guard str.count == 6 || str.count == 8 else { return .white }
        var rgba: UInt64 = 0
        Scanner(string: str).scanHexInt64(&rgba)
        let r, g, b, a: UInt64
        if str.count == 8 {
            a = (rgba & 0xFF000000) >> 24
            r = (rgba & 0x00FF0000) >> 16
            g = (rgba & 0x0000FF00) >> 8
            b = (rgba & 0x000000FF)
        } else {
            a = 255
            r = (rgba & 0xFF0000) >> 16
            g = (rgba & 0x00FF00) >> 8
            b = (rgba & 0x0000FF)
        }
        return UIColor(red: CGFloat(r) / 255.0,
                       green: CGFloat(g) / 255.0,
                       blue: CGFloat(b) / 255.0,
                       alpha: CGFloat(a) / 255.0)
    }
}

// MARK: - ChartViewDelegate
extension EcChartsView: ChartViewDelegate { }